client/allocrunner/alloc_runner.go

```diff
@@ -347,12 +347,26 @@ func (ar *allocRunner) shouldRun() bool {
 
 // runTasks is used to run the task runners and block until they exit.
 func (ar *allocRunner) runTasks() {
+	// Start all tasks
 	for _, task := range ar.tasks {
 		go task.Run()
 	}
 
+	// Block on all tasks except poststop tasks
 	for _, task := range ar.tasks {
-		<-task.WaitCh()
+		if !task.IsPoststopTask() {
+			<-task.WaitCh()
+		}
+	}
+
+	// Signal poststop tasks to proceed to main runtime
+	ar.taskHookCoordinator.StartPoststopTasks()
+
+	// Wait for poststop tasks to finish before proceeding
+	for _, task := range ar.tasks {
+		if task.IsPoststopTask() {
+			<-task.WaitCh()
+		}
```

Comment:
This logic is responsible for starting poststop tasks after the other tasks have finished. 

This covers the `nomad job stop` case: 
- wait until existing tasks have finished running
- start poststop tasks
- wait until poststop tasks finish

Reply:
How does this compare to having the coordinator tracking main task completion?  The coordinator logic now already tracks task completions (so main tasks start after pre-start) - can we reuse the same state machine rather than add the logic here?

Reply:
This was discussed offline & the goal is to do a refactor in another pull request to move the poststop specific code out of the alloc runner & back into the task hook coordinator. The reason it was avoided this time around because I didn't know how to block progression of the allocrunner from exiting from within the task hook coordinator.

In hindsight, one approach (instead of the for loop that blocks on poststop tasks) we could replace this loop with a channel that the TaskHookCoordinator closes when all the poststop tasks had finish. This would require introducing another set to track poststop task state.
```
<- ar.taskHookCoordinator.PostStopHooksFinished()
```